\documentclass[a4paper]{article}

\usepackage{listings}
\usepackage{indentfirst}

\title{RareVariantVis 2: R suite for analysis of rare variants in whole genome sequencing data.}
\author{Adam Gudys and Tomasz Stokowy}

%\VignetteIndexEntry{RareVariantVis}
%\VignetteEngine{knitr::knitr}

\begin{document}

\maketitle

\section*{Introduction}

The search for causative genetic variants in rare diseases of presumed monogenic inheritance has been boosted by the implementation of whole genome sequencing(WGS). Analysis and visualisation of WGS data is demanding due to its size and complexity.
To aid this challenge, we have developed a WGS data analysis suite â€“ RareVariantVis 2. This new, significantly extended implementation of RareVariantVis (Stokowy et al., Bioinformatics 2016) annotates, filters and visualises whole human genome in less than 30 minutes. Method accepts and integrates vcf files for single nucleotide, structural and copy number variants. Proposed method was successfully used to disclose causes of three rare monogenic disorders, including one non-coding variant.

This vignette was created to present how to efficiently visualize and interprete genomic variants in R. Package RareVariantVis aims to present genomic variants (especially rare ones) in a global, per chromosome way. Visualization is performed in two ways---standard that outputs png figures and interactive that uses JavaScript d3 package. Interactive visualization allows to analyze trio/family data, for example in search for causative variants in rare Mendelian diseases.

This vignette presents example of Genome in a Bottle sample - Ashkenazim trio son (chromosome 19) which is analyzed in about one minute on laptop/desktop computer. Examples include reading of all chr19 variants, filtering, annotation, visualization and homozygous region calling.

\section*{Example}
The analysis is preceeded by loading neccessary packages.
<<mychunk1, cache=TRUE, eval=TRUE, hide=TRUE>>=
library(RareVariantVis)
library(DataRareVariantVis)
@
\noindent
After that, single single nucleotide variants (SNV) and structural variants (SV) for Ashenazim trio son have to be loaded from \texttt{DataRareVariantVis} package.
<<mychunk2, cache=TRUE, eval=TRUE, hide=TRUE>>=
sample = system.file("extdata", "CoriellIndex_S1.vcf.gz",
    package = "DataRareVariantVis")
sv_sample = system.file("extdata", "CoriellIndex_S1.sv.vcf.gz",
    package = "DataRareVariantVis")
@
\noindent
The \texttt{RareVariantVis 2} package supports vcf files prepared by GATK variant caller. The discovery of rare variants in chromosome 19 is performed by executing the following command.
<<mychunk3, cache=TRUE, eval=TRUE, hide=TRUE>>=
chromosomeVis(sample=sample, sv_sample=sv_sample,
    chromosomes=c("19"))
@
\noindent
As a result, the package creates a set of files:
\begin{itemize}
    \item \textit{RareVariants\_CoriellIndex\_S1.txt}---table with all discovered rare single nucleotide variants,
    \item \textit{ComplexVariants\_CoriellIndex\_S1.txt}---table with all discovered rare single nucleotide variants,
    \item \textit{StructuralVariants\_CoriellIndex\_S1.txt}---table with all discovered rare structural variants,
    \item \textit{CoriellIndex\_S1\_chr19.png}---per-chromosome visualisation of discovered variants.
\end{itemize}
\noindent
If more than one chromosome is to be processed, separate visualisation files are generated. The table with rare variants description has following columns:
\begin{itemize}
    \item chromosome---number of chromosome,
    \item ...
\end{itemize}


\end{document}
